File: mascii2/src/main/java/com/kastkode/mascii2/musicelements/KeySignature.java
package com.kastkode.mascii2.musicelements;

import com.kastkode.mascii2.util.MasciiUtil;

public enum KeySignature {
	C_FLAT('c', Accidental.FLAT, setFlats('b', 'e', 'a', 'd', 'g', 'c', 'f'), -7),
	G_FLAT('g', Accidental.FLAT, setFlats('b', 'e', 'a', 'd', 'g', 'c'), -6),
	D_FLAT('d', Accidental.FLAT, setFlats('b', 'e', 'a', 'd', 'g'), -5),
	A_FLAT('a', Accidental.FLAT, setFlats('b', 'e', 'a', 'd'), -4),
	E_FLAT('e', Accidental.FLAT, setFlats('b', 'e', 'a'), -3),
	B_FLAT('b', Accidental.FLAT, setFlats('b', 'e'), -2),
	F('f', Accidental.FLAT, setFlats('b'), -1),
	C('c', Accidental.NATURAL, new AdHoc(), 0),
	G('g', Accidental.SHARP, setSharps('f'), 1),
	D('d', Accidental.SHARP, setSharps('f', 'c'), 2),
	A('a', Accidental.SHARP, setSharps('f', 'c', 'g'), 3),
	E('e', Accidental.SHARP, setSharps('f', 'c', 'g', 'd'), 4),
	B('b', Accidental.SHARP, setSharps('f', 'c', 'g', 'd', 'a'), 5),
	F_SHARP('f', Accidental.SHARP, setSharps('f', 'c', 'g', 'd', 'a', 'e'), 6),
	C_SHARP('c', Accidental.SHARP, setSharps('f', 'c', 'g', 'd', 'a', 'e', 'b'), 7);

	//TODO: calculate relative minor.  what abt A# minor?
	
	private static KeySignature[] flatSignatures = new KeySignature[] {
			A_FLAT, B_FLAT, null, D_FLAT, E_FLAT, null, G_FLAT
	};
	private static KeySignature[] naturalSignatures = new KeySignature[] {
			A, B, C, D, E, F, G
	};
	private static KeySignature[] sharpSignatures = new KeySignature[] {
			null, null, C_SHARP, null, null, F_SHARP, null
	};
	
	private char degree;
	private Accidental accidental;
	private int code;
	private AdHoc alterations;
	
	KeySignature(char degree, Accidental accidental, AdHoc alterations, int code) {
		this.degree = degree;
		this.accidental = accidental;
		this.alterations = alterations;
		this.code = code;
	}
	
	public static KeySignature getByCode(int code) {
		for(KeySignature k:values()) {
			if(k.code == code) return k;
		}
		return null;
	}

	public static KeySignature getByDegree(char degree, Accidental accidental, boolean isMinor) {
		degree = Character.toLowerCase(degree);
		
		
		if(! isMinor) {
			KeySignature[] keys;
			if(accidental == Accidental.FLAT) {
				keys = flatSignatures;
			} else if(accidental == Accidental.SHARP) {
				keys = sharpSignatures;
			} else {
				keys = naturalSignatures;
			}
			
			int majindex = (degree - 'a') % keys.length;
			KeySignature maj = keys[majindex];
			return maj;
		} else {
			//TODO: there must be a better way... but meanwhile this works
			
			//1. figure out the 6th note in this scale
			int relMaj = (MasciiUtil.relativeMajorDegree(degree) - 'a') % naturalSignatures.length;
			
			if(naturalSignatures[relMaj].alterations.getAccidental(degree) == accidental) {
				return naturalSignatures[relMaj];
			} else if(sharpSignatures[relMaj] != null && sharpSignatures[relMaj].alterations.getAccidental(degree) == accidental) {
				return sharpSignatures[relMaj];
			} else {
				return flatSignatures[relMaj];
			}
		}
				
	}
	
	public char getDegree() {
		return degree;
	}

	public Accidental getAccidental() {
		return accidental;
	}

	//always returns defensive copy, not original
	public AdHoc alterationsMap() {
		return alterations.copy();
	}
	
	public int getCode() {
		return code;
	}

	private static final AdHoc setFlats(char... degrees) {
		AdHoc answer = new AdHoc();
		for(char ch:degrees) {
			answer.setAccidental(ch, Accidental.FLAT);
		}
		return answer;
	}
	
	private static final AdHoc setSharps(char... degrees) {
		AdHoc answer = new AdHoc();
		for(char c:degrees) {
			answer.setAccidental(c, Accidental.SHARP);
		}
		return answer;
	}
	
	
	public static class AdHoc {
		private static final int SHARPS_BYTE = 0 * 8; //last byte, byte 0
		private static final int DBLSHARPS_BYTE = 1 * 8; //byte 1
		private static final int FLATS_BYTE = 2 * 8; //byte 2
		private static final int DBLFLATS_BYTE = 3 * 8; //byte 3
		
		private int accidentalsBitmap;

		public AdHoc() {
			this(0);
		}
		
		public AdHoc(int accidentalsBitmap) {
			this.accidentalsBitmap = accidentalsBitmap;
		}
		
		public AdHoc copy() {
			return new AdHoc(this.accidentalsBitmap);
		}
		
		public int asBitmap() {
			return accidentalsBitmap;
		}

		public Accidental getAccidental(char c) {
			c = Character.toLowerCase(c);
			int offset = c - 'a';
			
			if ((accidentalsBitmap & (1 << offset + DBLSHARPS_BYTE)) != 0)
			{
				return Accidental.DBLSHARP;
			} else if ((accidentalsBitmap & (1 << offset + DBLFLATS_BYTE)) != 0)
			{
				return Accidental.DBLFLAT;
			} else if((accidentalsBitmap & (1 << offset + SHARPS_BYTE)) != 0)
			{
				return Accidental.SHARP;
			} else if((accidentalsBitmap & (1 << offset + FLATS_BYTE)) != 0)
			{
				return Accidental.FLAT;
			} else {
				return Accidental.NATURAL;
			}
		}
		
		public int setAccidental(char c, Accidental acc) {
			c = Character.toLowerCase(c);
			int offset = c - 'a';

			if(acc == Accidental.DBLFLAT) {
				accidentalsBitmap |= 1 << (offset + DBLFLATS_BYTE);			
			} else {
				accidentalsBitmap &= ~(1 << (offset + DBLFLATS_BYTE));					
			}

			if(acc == Accidental.DBLSHARP) {
				accidentalsBitmap |= 1 << (offset + DBLSHARPS_BYTE);			
			} else {
				accidentalsBitmap &= ~(1 << (offset + DBLSHARPS_BYTE));					
			}

			if(acc == Accidental.FLAT) {
				accidentalsBitmap |= 1 << (offset + FLATS_BYTE);			
			} else {
				accidentalsBitmap &= ~(1 << (offset + FLATS_BYTE));					
			}

			if(acc == Accidental.SHARP) {
				accidentalsBitmap |= 1 << (offset + SHARPS_BYTE);			
			} else {
				accidentalsBitmap &= ~(1 << (offset + SHARPS_BYTE));					
			}
			return accidentalsBitmap;
		}

	}
	
	@Override
	public String toString() {
		return code + "";
	}
}

File: mascii2-typescript/src/musicelements/KeySignature.ts
import { Accidental } from './Accidental';

export class KeySignatureAdHoc {
    private static readonly SHARPS_BYTE = 0 * 8;
    private static readonly DBLSHARPS_BYTE = 1 * 8;
    private static readonly FLATS_BYTE = 2 * 8;
    private static readonly DBLFLATS_BYTE = 3 * 8;

    private bitmap: number;

    constructor(bitmap = 0) {
        this.bitmap = bitmap;
    }

    copy(): KeySignatureAdHoc {
        return new KeySignatureAdHoc(this.bitmap);
    }

    asBitmap(): number {
        return this.bitmap;
    }

    getAccidental(c: string): Accidental {
        const offset = c.toLowerCase().charCodeAt(0) - 'a'.charCodeAt(0);
        if ((this.bitmap & (1 << (offset + KeySignatureAdHoc.DBLSHARPS_BYTE))) !== 0) return Accidental.DBLSHARP;
        if ((this.bitmap & (1 << (offset + KeySignatureAdHoc.DBLFLATS_BYTE))) !== 0) return Accidental.DBLFLAT;
        if ((this.bitmap & (1 << (offset + KeySignatureAdHoc.SHARPS_BYTE))) !== 0) return Accidental.SHARP;
        if ((this.bitmap & (1 << (offset + KeySignatureAdHoc.FLATS_BYTE))) !== 0) return Accidental.FLAT;
        return Accidental.NATURAL;
    }

    setAccidental(c: string, acc: Accidental): number {
        const offset = c.toLowerCase().charCodeAt(0) - 'a'.charCodeAt(0);
        if (acc === Accidental.DBLFLAT) {
            this.bitmap |= 1 << (offset + KeySignatureAdHoc.DBLFLATS_BYTE);
        } else {
            this.bitmap &= ~(1 << (offset + KeySignatureAdHoc.DBLFLATS_BYTE));
        }
        if (acc === Accidental.DBLSHARP) {
            this.bitmap |= 1 << (offset + KeySignatureAdHoc.DBLSHARPS_BYTE);
        } else {
            this.bitmap &= ~(1 << (offset + KeySignatureAdHoc.DBLSHARPS_BYTE));
        }
        if (acc === Accidental.FLAT) {
            this.bitmap |= 1 << (offset + KeySignatureAdHoc.FLATS_BYTE);
        } else {
            this.bitmap &= ~(1 << (offset + KeySignatureAdHoc.FLATS_BYTE));
        }
        if (acc === Accidental.SHARP) {
            this.bitmap |= 1 << (offset + KeySignatureAdHoc.SHARPS_BYTE);
        } else {
            this.bitmap &= ~(1 << (offset + KeySignatureAdHoc.SHARPS_BYTE));
        }
        return this.bitmap;
    }
}

export class KeySignature {
    readonly degree: string;
    readonly accidental: Accidental;
    readonly code: number;
    private readonly _alterations: KeySignatureAdHoc;

    private constructor(degree: string, accidental: Accidental, alterations: KeySignatureAdHoc, code: number) {
        this.degree = degree;
        this.accidental = accidental;
        this._alterations = alterations;
        this.code = code;
    }

    alterationsMap(): KeySignatureAdHoc {
        return this._alterations.copy();
    }

    toString(): string {
        return `${this.code}`;
    }

    private static setFlats(...degrees: string[]): KeySignatureAdHoc {
        const ans = new KeySignatureAdHoc();
        for (const ch of degrees) ans.setAccidental(ch, Accidental.FLAT);
        return ans;
    }

    private static setSharps(...degrees: string[]): KeySignatureAdHoc {
        const ans = new KeySignatureAdHoc();
        for (const ch of degrees) ans.setAccidental(ch, Accidental.SHARP);
        return ans;
    }

    static readonly C_FLAT  = new KeySignature('c', Accidental.FLAT,    KeySignature.setFlats('b','e','a','d','g','c','f'), -7);
    static readonly G_FLAT  = new KeySignature('g', Accidental.FLAT,    KeySignature.setFlats('b','e','a','d','g','c'),     -6);
    static readonly D_FLAT  = new KeySignature('d', Accidental.FLAT,    KeySignature.setFlats('b','e','a','d','g'),         -5);
    static readonly A_FLAT  = new KeySignature('a', Accidental.FLAT,    KeySignature.setFlats('b','e','a','d'),             -4);
    static readonly E_FLAT  = new KeySignature('e', Accidental.FLAT,    KeySignature.setFlats('b','e','a'),                 -3);
    static readonly B_FLAT  = new KeySignature('b', Accidental.FLAT,    KeySignature.setFlats('b','e'),                     -2);
    static readonly F       = new KeySignature('f', Accidental.FLAT,    KeySignature.setFlats('b'),                         -1);
    static readonly C       = new KeySignature('c', Accidental.NATURAL, new KeySignatureAdHoc(),                             0);
    static readonly G       = new KeySignature('g', Accidental.SHARP,   KeySignature.setSharps('f'),                         1);
    static readonly D       = new KeySignature('d', Accidental.SHARP,   KeySignature.setSharps('f','c'),                     2);
    static readonly A       = new KeySignature('a', Accidental.SHARP,   KeySignature.setSharps('f','c','g'),                 3);
    static readonly E       = new KeySignature('e', Accidental.SHARP,   KeySignature.setSharps('f','c','g','d'),             4);
    static readonly B       = new KeySignature('b', Accidental.SHARP,   KeySignature.setSharps('f','c','g','d','a'),         5);
    static readonly F_SHARP = new KeySignature('f', Accidental.SHARP,   KeySignature.setSharps('f','c','g','d','a','e'),     6);
    static readonly C_SHARP = new KeySignature('c', Accidental.SHARP,   KeySignature.setSharps('f','c','g','d','a','e','b'), 7);

    private static readonly allKeys: KeySignature[] = [
        KeySignature.C_FLAT, KeySignature.G_FLAT, KeySignature.D_FLAT, KeySignature.A_FLAT,
        KeySignature.E_FLAT, KeySignature.B_FLAT, KeySignature.F,      KeySignature.C,
        KeySignature.G,      KeySignature.D,      KeySignature.A,      KeySignature.E,
        KeySignature.B,      KeySignature.F_SHARP, KeySignature.C_SHARP,
    ];

    // Indexed [a, b, c, d, e, f, g] (offset from 'a')
    private static readonly flatSignatures: (KeySignature | null)[] = [
        KeySignature.A_FLAT, KeySignature.B_FLAT, null, KeySignature.D_FLAT,
        KeySignature.E_FLAT, null, KeySignature.G_FLAT,
    ];
    private static readonly naturalSignatures: KeySignature[] = [
        KeySignature.A, KeySignature.B, KeySignature.C, KeySignature.D,
        KeySignature.E, KeySignature.F, KeySignature.G,
    ];
    private static readonly sharpSignatures: (KeySignature | null)[] = [
        null, null, KeySignature.C_SHARP, null, null, KeySignature.F_SHARP, null,
    ];

    static getByCode(code: number): KeySignature | null {
        return KeySignature.allKeys.find(k => k.code === code) ?? null;
    }

    static getByDegree(degree: string, accidental: Accidental, isMinor: boolean): KeySignature | null {
        const d = degree.toLowerCase();
        if (!isMinor) {
            let keys: (KeySignature | null)[];
            if (accidental === Accidental.FLAT) keys = KeySignature.flatSignatures;
            else if (accidental === Accidental.SHARP) keys = KeySignature.sharpSignatures;
            else keys = KeySignature.naturalSignatures;
            const idx = (d.charCodeAt(0) - 'a'.charCodeAt(0)) % keys.length;
            return keys[idx] ?? null;
        } else {
            // relativeMajorDegree: (degree + 2) % 7 (gives the 6th note going backwards = relative major)
            const relMaj = ((d.charCodeAt(0) - 'a'.charCodeAt(0)) + 2) % 7;
            const nat = KeySignature.naturalSignatures[relMaj];
            if (nat && nat._alterations.getAccidental(d) === accidental) return nat;
            const sharp = KeySignature.sharpSignatures[relMaj];
            if (sharp && sharp._alterations.getAccidental(d) === accidental) return sharp;
            return KeySignature.flatSignatures[relMaj] ?? null;
        }
    }
}
